storage-proofs-porep/src/stacked/vanilla/proof.rs

```diff
@@ -105,66 +105,107 @@ impl<'a, Tree: 'static + MerkleTreeTrait, G: 'static + Hasher> StackedDrg<'a, Tr
-            );
-        }
 
-        // If synthetic vanilla proofs are stored on disk, read and return the proofs corresponding
-        // to the porep challlenge set.
-        let read_synth_proofs = layer_challenges.use_synthetic && pub_inputs.seed.is_some();
-        if read_synth_proofs {
-            let read_res = Self::read_porep_proofs_from_synth(
-                graph_size,
-                pub_inputs,
-                layer_challenges,
-                t_aux,
-                partition_count,
-            );
-            if read_res.is_ok() {
-                return read_res;
-            }
-            info!(
-                "failed to read porep proofs from synthetic proofs file: {:?}",
-                t_aux.synth_proofs_path(),
-            );
+        if layer_challenges.use_synthetic {
+            // If there are no synthetic vanilla proofs stored on disk yet, generate them.
+            if pub_inputs.seed.is_none() {
+                info!("generating synthetic vanilla proofs in a single partition");
+                assert_eq!(partition_count, 1);
 
-            // If the synthetic proofs file does not exist and we have the layers available,
-            // we can generate non-synthetic proofs
-            if t_aux.labels.len() == layers {
-                info!("skipping synthetic proving; generating non-synthetic vanilla proofs");
-            } else {
-                error!("synthetic proving failure; synthetic proofs and layers are unavailable");
+                let comm_r = pub_inputs.tau.as_ref().expect("tau is set").comm_r;
+                // Derive the set of challenges we are proving over.
+                let challenges = layer_challenges.derive_synthetic(
+                    graph.size(),
+                    &pub_inputs.replica_id,
+                    &comm_r,
+                );
+
+                let synth_proofs = Self::prove_layers_generate(
+                    graph,
+                    pub_inputs,
+                    p_aux.comm_c,
+                    t_aux,
+                    challenges,
+                    layers,
+                )?;
 
-                return read_res;
+                Self::write_synth_proofs(
+                    &synth_proofs,
+                    pub_inputs,
+                    graph,
+                    layer_challenges,
+                    t_aux.synth_proofs_path(),
+                )?;
+                Ok(vec![vec![]; partition_count])
             }
-        }
+            // Else the synthetic vanilla proofs are stored on disk, read and return the proofs
+            // corresponding to the porep challlenge set.
+            else {
+                Self::read_porep_proofs_from_synth(
+                    graph.size(),
+                    pub_inputs,
+                    layer_challenges,
+                    t_aux.synth_proofs_path(),
+                    partition_count,
+                )
+                .map_err(|error| {
```

Comment:
In the case where Synth-PoRep is enabled and the challenge seed is set, but synthetic challenge proofs do not exist on disk, I think the fallback behavior should be to run normal PoRep (rather than return an error). In the event where, while waiting for the challenge seed on chain, a user's synthetic challenge proofs were moved/removed from their expected location, returning an error here would run in an endless loop (assuming the user were to rerun `prove_layers` following the read synth-proofs error).

Reply:
If a user runs a Synthetic PoRep, you would do that in order to free space. So you would remove the tree files/layers once the proofs are generated. If you move/remove those proofs and fallback to non-Synthetic PoRep, it would likely fail, because the original files are not there anymore. I would find that behaviour more confusing/hard to follow that failing early on the Synthetic PoRep code path.

This is kind of a discussion, whether we want "fail early" behaviour, which I prefer as I find it easier to debug/finding root cases if it happend. Vs. "best effort", which might then make things fail later on in unexpected ways.

Reply:
Yeah, I think that it's totally debatable, and that the error makes sense too. Do you you have an opinion about falling back to standard PoRep only when the SDR layers weren't deleted? Or do you think an error should be returned in that case too?

Reply:
> Do you you have an opinion about falling back to standard PoRep only when the SDR layers weren't deleted? Or do you think an error should be returned in that case too?

That might be a good middle ground. Though if it would be me, I think would still return early as I really see it as some misconfiguration/user error which under normal operations should never happen. And If someone did something wrong, they probably want to find out about it early. I'm also thinking about potential bug reports, which might be easier to tackle with early failure.

Reply:
I'm ok with erroring as it would force a user to immediately diagnose and fix the issue. My thinking for why standard PoRep should be run if the labels are still present on disk (with a logged warning saying "synth-porep was not run") is mainly because I think that re-sealing/labeling should be avoided when possible. However, if that fallback behavior is repeatedly taken, a user may be mislead into thinking that they are using synth-porep when in reality they aren't.

Maybe @cryptonemo has an opinion, but if not, I'm happy to approve.

Reply:
My thought on it is that we should fallback to PoRep if it's possible and warn/log very loudly that something is misconfigured and SynthPoRep cannot be used.  I agree that will complicate potential code/testing paths, but shouldn't complicate much past having the user reading the logs and clearly seeing there's a problem.

Maybe the more ideal answer from a user perspective is to check if the synth_proofs data exists *in the clear data call (when enabled of course)* and log loudly there that they cannot be cleared because synth_proofs data doesn't exist -- and then we have the logic to fall back to PoRep because something is wildly misconfigured ???

Reply:
> I agree that will complicate potential code/testing paths, but shouldn't complicate much past having the user reading the logs and clearly seeing there's a problem.

I disagree here. Especially with Ni-PoRep there's yet another code path. I find the current code pretty clean, you can clearly see the different PoRep cases and reason about what happens when. With certain fallback it likely be a lot of additional back and forth, which I think is just not worth it.

From a users perspective I see it as: if i've misconfigured wildly I want to immediately know it, rather then having a system that tries random things just in hope it might recover somehow, but might still end up in a broken state. An example for today: SupraSeal's PC2 does not support Synthetic PoRep. When run with Lotus and SyntheticPoRep enabled, it runs into a seemingly endless loop, but then does something (the wrong thing). Wouldn't it be much better if it would just say: "error, you try to run Synthetic PoRep although it's not supported"?

So if there's *any* chance you could live with the current solution I'd prefer keeping it the current way.

If you can't live with it, should we then also start thinking about how it will look like if we also add Ni-PoRep. Should there also be fallbacks?

Reply:
Ok, that's fair.  Thanks!